xarray/core/common.py

```diff
@@ -479,8 +482,37 @@ def where(self, cond):
         Coordinates:
           * y        (y) int64 0 1 2 3 4
           * x        (x) int64 0 1 2 3 4
+        >>> a.where((a > 6) & (a < 18), drop=True)
+        <xarray.DataArray (x: 5, y: 5)>
+        array([[ nan,  nan,   7.,   8.,   9.],
+               [ 10.,  11.,  12.,  13.,  14.],
+               [ 15.,  16.,  17.,  nan,  nan],
+        Coordinates:
+          * x        (x) int64 1 2 3
+          * y        (y) int64 0 1 2 3 4
         """
-        return self._where(cond)
+        if drop:
+            from .dataarray import DataArray
+            from .dataset import Dataset
+            # get cond with the minimal size needed for the Dataset
+            if isinstance(cond, Dataset):
+                clipcond = cond.to_array().any('variable')
+            elif isinstance(cond, DataArray):
+                clipcond = cond
+            else:
+                raise ValueError("Cond argument is %r but must be a %r or %r" %
+                                 (cond, Dataset, DataArray))
+
+            # clip the data corresponding to coordinate dims that are not used
+            clip = dict(zip(clipcond.dims,[np.unique(adim) for adim in np.nonzero(clipcond.values)]))
```

Comment:
this line is getting a little long -- move the list comprehension to its own line?